### PR TITLE
Uber - Use bearer token

### LINKED ISSUE #114: Uber API - Replace server tokens with bearer tokens
## Expected behavior

The Uber API should be called using bearer tokens.
The Uber API is phasing out server tokens currently used in this OTP fork.

## Observed behavior

The Uber API is being called using server tokens that are being phased out.

## Version of OTP used (exact commit hash or JAR name)

Any recent ibi-dev commit.

## Data sets in use (links to GTFS and OSM PBF files)

NA

## Command line used to start OTP

NA

## Router config and graph build config JSON

Any router config that specifies a `serverToken` for the Uber API.

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
To be completed by pull request submitter:

- [x] **issue**: Fix #114.
- [na] **roadmap**: This change is requested by TriMet.
- [x] **tests**: Have you added relevant test coverage? => yes. Are all the tests passing on [the continuous integration service (Travis CI)](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#continuous-integration)? => yes
- [x] **formatting**: Have you followed the [suggested code style](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Developers-Guide.md#code-style)? 
- [x] **documentation**: If you are adding a new configuration option, have you added an explanation to the [configuration documentation](docs/Configuration.md) tables and sections? => yes
- [na] **changelog**: (does not apply to this fork) add a bullet point to the [changelog file](https://github.com/opentripplanner/OpenTripPlanner/blob/master/docs/Changelog.md) with description and link to the linked issue

To be completed by @opentripplanner/plc:

- [na] reviews and approvals by 2 members, ideally from different organizations
- [na] **after merging**: update the relevant card on the [roadmap](https://github.com/orgs/opentripplanner/projects/1)

### CODE REVIEW

**Review by demory — APPROVED** [review]

File: src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import java.net.HttpURLConnection;
+import java.time.Instant;
+import java.util.Date;
+
+/**
+ * Holds an OAuth access token and its expiration time for querying ride-hail APIs.
+ */
+public class OAuthToken {
+    public final String value;
+    private Date tokenExpirationTime;
```
Comment (leonardehrenfried) [inline src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java]: I'd recommend to use something from the `java.time` package, for example `Instant`. `java.util.Date` has so many pitfalls!

File: src/main/java/org/opentripplanner/updater/transportation_network_company/lyft/LyftTransportationNetworkCompanyDataSource.java [hunk 2]
```diff
@@ -61,51 +58,35 @@ public LyftTransportationNetworkCompanyDataSource(String baseUrl, String clientI
...
-    }
-
     @Override
     public TransportationNetworkCompany getTransportationNetworkCompanyType() {
         return TransportationNetworkCompany.LYFT;
     }
 
+    @Override
+    protected HttpURLConnection buildOAuthConnection() throws IOException {
+        ObjectMapper mapper = new ObjectMapper();
```
Comment (leonardehrenfried) [inline src/main/java/org/opentripplanner/updater/transportation_network_company/lyft/LyftTransportationNetworkCompanyDataSource.java]: The mapper can be a static field.

File: src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        value = null;
+    }
+
+    public static OAuthToken blank() {
+        return new OAuthToken();
+    }
+
+    public OAuthToken(HttpURLConnection connection) {
+        // send request and parse response
+        ObjectMapper mapper = new ObjectMapper();
```
Comment (leonardehrenfried) [inline src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java]: Can be static.

File: src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        String newValue = null;
+        try (InputStream responseStream = connection.getInputStream()) {
+            OAuthAuthenticationResponse response = mapper.readValue(responseStream, OAuthAuthenticationResponse.class);
+
+            newValue = response.access_token;
+            tokenExpirationTime = new Date();
+            tokenExpirationTime.setTime(tokenExpirationTime.getTime() + (response.expires_in - 60) * 1000L);
+        } catch (IOException e) {
+
```
Comment (leonardehrenfried) [inline src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java]: Please do something with the exception even if you're only logging or rethrowing it.

**Review by leonardehrenfried — CHANGES_REQUESTED** [review]
I've picked up on a couple of things.

File: src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java (hunk 1 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java]: Thanks, I have updated in https://github.com/ibi-group/OpenTripPlanner/pull/115/commits/4b7970237b804bd0c64a14a9b7acda08302567be, it's much cleaner that way too.

File: src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java (hunk 4 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java]: Rethrowing the exception in https://github.com/ibi-group/OpenTripPlanner/pull/115/commits/4b7970237b804bd0c64a14a9b7acda08302567be.

File: src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java (hunk 3 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/updater/transportation_network_company/OAuthToken.java]: Updated in https://github.com/ibi-group/OpenTripPlanner/pull/115/commits/4b7970237b804bd0c64a14a9b7acda08302567be.

File: src/main/java/org/opentripplanner/updater/transportation_network_company/lyft/LyftTransportationNetworkCompanyDataSource.java (hunk 2 above)
Comment (binh-dam-ibigroup) [reply src/main/java/org/opentripplanner/updater/transportation_network_company/lyft/LyftTransportationNetworkCompanyDataSource.java]: Updated in https://github.com/ibi-group/OpenTripPlanner/pull/115/commits/0db7e55f333a0d43560d81900de07bcf56562db7, and for Uber too.

**Review by leonardehrenfried — APPROVED** [review]

**Pull request merged**